test/mongoose_push_test.exs

```diff
@@ -1,8 +1,152 @@
 defmodule MongoosePushTest do
   use ExUnit.Case
+  import MongoosePush
+  import Mock
   doctest MongoosePush
 
-  test "the truth" do
-    assert 1 + 1 == 2
+  setup do
+    reset(:fcm)
+    reset(:apns)
+  end
+
+  test "simple push to apns succeeds" do
+    assert :ok == push("device_id",
+                       %{:service => :apns, :title => "", :body => ""})
+  end
+
+  test "push to apns assign correct message fields" do
+    notification =
+      %{:service => :apns,
+        :title => "title value",
+        :body => "body value",
+        :badge => 5,
+        :click_action => "click.action"
+      }
+
+    assert :ok == push("testdeviceid1234", notification)
+
+    apns_request = last_activity(:apns)
+    aps_data = apns_request["request_data"]["aps"]
+
+    assert "testdeviceid1234" == apns_request["device_token"]
+    assert notification[:title] == aps_data["alert"]["title"]
+    assert notification[:body] == aps_data["alert"]["body"]
+    assert notification[:badge] == aps_data["badge"]
+    assert notification[:click_action] == aps_data["category"]
+
+  end
+
+  test "push to fcm assign correct message fields" do
+    notification =
+      %{:service => :fcm,
+        :title => "title value",
+        :body => "body value",
+        :click_action => "click.action",
+        :tag => "tag value"
+      }
+
+    assert :ok == push("androidtestdeviceid12", notification)
+    fcm_request = last_activity(:fcm)
+    fcm_data = fcm_request["request_data"]["notification"]
+
+    assert "androidtestdeviceid12" == fcm_request["device_token"]
+    assert notification[:title] == fcm_data["title"]
+    assert notification[:body] == fcm_data["body"]
+    assert notification[:click_action] == fcm_data["click_action"]
+    assert notification[:tag] == fcm_data["tag"]
+
+  end
+
+  test "push to fcm with unknown token fails" do
+    notification =
+      %{:service => :fcm,
+        :title => "title value",
+        :body => "body value",
+        :click_action => "click.action",
+        :tag => "tag value"
+      }
+    fail_tokens(:fcm, [%{device_token: "androidtestdeviceid65", status: 200,
+                         reason: "InvalidRegistration"}])
+
+    assert {:error, _} = push("androidtestdeviceid65", notification)
+  end
+
+  defp reset(service) do
```

Comment:
hmm, looks like the second clause of this function, which starts on line 107, has the same body.